File: Q13 three sum.js
function threeSum(nums) {
    const target = 0;
    nums.sort((a, b) => a - b);
    const uniqueTriplets = new Set();
    const output = [];

    for (let i = 0; i < nums.length; i++) {
        let j = i + 1;
        let k = nums.length - 1;
        while (j < k) {
            const sum = nums[i] + nums[j] + nums[k];
            if (sum === target) {
                uniqueTriplets.add(JSON.stringify([nums[i], nums[j], nums[k]]));
                j++;
                k--;
            } else if (sum < target) {
                j++;
            } else {
                k--;
            }
        }
    }

    uniqueTriplets.forEach(triplet => {
        output.push(JSON.parse(triplet));
    });

    return output;
}

File: Q13 three sum.py
class Solution:
    def threeSum(self, nums):
        target = 0
        nums.sort()
        unique_triplets = set()
        output = []

        for i in range(len(nums)):
            j = i + 1
            k = len(nums) - 1
            while j < k:
                sum_triplet = nums[i] + nums[j] + nums[k]
                if sum_triplet == target:
                    unique_triplets.add((nums[i], nums[j], nums[k]))
                    j += 1
                    k -= 1
                elif sum_triplet < target:
                    j += 1
                else:
                    k -= 1

        for triplet in unique_triplets:
            output.append(list(triplet))

        return output


